### PR TITLE
Fix/change hubspot api auth

### PR DESCRIPTION
**Pull request by npracht**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | ✅ 
| New feature/enhancement? (use the a.x branch)      | [
| Deprecations?                          | 
| BC breaks? (use the c.x branch)        | 
| Automated tests included?              | 
| Related user documentation PR URL      | https://github.com/mautic/mautic-documentation/pull/333
| Related developer documentation PR URL | 
| Issue(s) addressed                     | Fixes #... 

#### Description:
We need to change authentication method for Hubspot plugin. **New installation cannot use our plugin anymore**, and existing in integration will shut down in November, 22'.

Source: [developers.hubspot.com/changelog/upcoming-api-key-sunset](https://developers.hubspot.com/changelog/upcoming-api-key-sunset)

> What’s changing?
> API Keys have been [one of three authentication methods supported by HubSpot](https://developers.hubspot.com/docs/api/intro-to-auth) APIs. However, as part of ongoing efforts to protect our customer's data, we will be sunsetting API Keys.
> 
> As a result of this change, integrations will instead be required to work with [Private Apps](https://developers.hubspot.com/docs/api/private-apps). Private Apps offer tighter security and allow more granular control over your integrations and account data than legacy API keys.
> 
> What this means for developers:
> With the introduction of Private Apps, users who previously developed on HubSpot and utilized API Keys will now be required to migrate existing integrations from using API Key authentication to using Private Apps instead. Instructions for how to migrate existing integrations can be found [here](https://developers.hubspot.com/docs/api/migrate-an-api-key-integration-to-a-private-app).

Migration doc: [developers.hubspot.com/docs/api/migrate-an-api-key-integration-to-a-private-app](https://developers.hubspot.com/docs/api/migrate-an-api-key-integration-to-a-private-app)

#### Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Create a private app in hubspot with following rights
3. Setup plugin integration
4. See everything works
image

### CODE REVIEW

**Review by escopecz — APPROVED** [review]
I see no issue with the code changes. Thanks! 👍

File: plugins/MauticCrmBundle/Integration/HubspotIntegration.php [hunk 1]
```diff
@@ -127,14 +128,32 @@ public function getSupportedFeatures()
...
 
+    public function getBearerToken($inAuthorization = false)
+    {
+        $tokenData = $this->getKeys();
+
+        return $tokenData[self::ACCESS_KEY] ?? null;
+    }
+
+    /**
+     * @return array
```
Comment (shinde-rahul) [inline plugins/MauticCrmBundle/Integration/HubspotIntegration.php]:
```suggestion
     * @return <string, bool>
```

**Review by shinde-rahul — CHANGES_REQUESTED** [review]

**Comment by RCheesley** [comment]
@npracht @escopecz please see suggestion above from @shinde-rahul

**Comment by RCheesley** [comment]
(also, PHPSTAN needs some ❤️ )

**Review by RCheesley — CHANGES_REQUESTED** [review]
When I try to set up the plugin configuration, there are no fields for adding the authentication credentials.

@npracht @kuzmany could you take a look?

I have tried clicking to install plugins again.

screenshot-8080-mautic-mautic-xjh7w94uz3a ws-eu67 gitpod io-2022 09 23-15_55_27

**Comment by kuzmany** [comment]
@RCheesley fixed

**Comment by RCheesley** [comment]
Fields now show but they are not editable:

screenshot-8080-mautic-mautic-5xm3ui5dau8 ws-eu67 gitpod io-2022 09 24-12_35_42

**Comment by kuzmany** [comment]
@RCheesley fixed

**Comment by RCheesley** [comment]
OK here's the outcome from my tests:

0. I was able to add the API key and save it, then to turn on the integration and set up the field mapping. 🥳 

1. Constantly getting the tooltip stuck on the browser:
screenshot-8080-mautic-mautic-swe95w1uz6p ws-eu67 gitpod io-2022 09 24-19_48_43

2. Only 8 contacts are synchronised with the cron job:
screenshot-app-eu1 hubspot com-2022 09 24-19_55_40

3. Creating a campaign action to push contacts to the integration throws an error when I select the action 'push contact to integration' so I was unable to test it (possibly related to https://github.com/mautic/mautic/issues/11245):

```
[2022-09-24 17:54:31] mautic.CRITICAL: Uncaught PHP Exception ErrorException: "PHP Notice - Trying to access array offset on value of type null" at /var/www/html/app/bundles/PluginBundle/Form/Type/IntegrationsListType.php line 102 {"exception":"[object] (ErrorException(code: 0): PHP Notice - Trying to access array offset on value of type null at /var/www/html/app/bundles/PluginBundle/Form/Type/IntegrationsListType.php:102)
[stacktrace]
#0 /var/www/html/app/bundles/PluginBundle/Form/Type/IntegrationsListType.php(102): Mautic\\CoreBundle\\ErrorHandler\\ErrorHandler->handleError(8, 'PHP Notice - Tr...', '/var/www/html/a...', 102, Array)
#1 /var/www/html/app/bundles/PluginBundle/Form/Type/IntegrationsListType.php(127): Mautic\\PluginBundle\\Form\\Type\\IntegrationsListType->Mautic\\PluginBundle\\Form\\Type\\{closure}(Object(Symfony\\Component\\Form\\Form), NULL)
#2 /var/www/html/vendor/symfony/event-dispatcher/EventDispatcher.php(264): Mautic\\PluginBundle\\Form\\Type\\IntegrationsListType->Mautic\\PluginBundle\\Form\\Type\\{closure}(Object(Symfony\\Component\\Form\\Event\\PreSetDataEvent), 'form.pre_set_da...', Object(Symfony\\Component\\EventDispatcher\\EventDispatcher))
#3 /var/www/html/vendor/symfony/event-dispatcher/EventDispatcher.php(239): Symfony\\Component\\EventDispatcher\\EventDispatcher->doDispatch(Array, 'form.pre_set_da...', Object(Symfony\\Component\\Form\\Event\\PreSetDataEvent))
#4 /var/www/html/vendor/symfony/event-dispatcher/EventDispatcher.php(73): Symfony\\Component\\EventDispatcher\\EventDispatcher->callListeners(Array, 'form.pre_set_da...', Object(Symfony\\Component\\Form\\Event\\PreSetDataEvent))
#5 /var/www/html/vendor/symfony/event-dispatcher/ImmutableEventDispatcher.php(44): Symfony\\Component\\EventDispatcher\\EventDispatcher->dispatch(Object(Symfony\\Component\\Form\\Event\\PreSetDataEvent), 'form.pre_set_da...')
#6 /var/www/html/vendor/symfony/form/Form.php(353): Symfony\\Component\\EventDispatcher\\ImmutableEventDispatcher->dispatch(Object(Symfony\\Component\\Form\\Event\\PreSetDataEvent), 'form.pre_set_da...')
#7 /var/www/html/vendor/symfony/form/Extension/Core/DataMapper/PropertyPathMapper.php(51): Symfony\\Component\\Form\\Form->setData(NULL)
#8 /var/www/html/vendor/symfony/form/Form.php(389): Symfony\\Component\\Form\\Extension\\Core\\DataMapper\\PropertyPathMapper->mapDataToForms(Array, Object(RecursiveIteratorIterator))
#9 /var/www/html/vendor/symfony/form/Form.php(493): Symfony\\Component\\Form\\Form->setData(Array)
#10 /var/www/html/vendor/symfony/form/FormBuilder.php(208): Symfony\\Component\\Form\\Form->initialize()
#11 /var/www/html/vendor/symfony/form/FormFactory.php(32): Symfony\\Component\\Form\\FormBuilder->getForm()
#12 /var/www/html/app/bundles/CampaignBundle/Controller/EventController.php(80): Symfony\\Component\\Form\\FormFactory->create('Mautic\\\\Campaign...', Array, Array)
#13 /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php(447): Mautic\\CampaignBundle\\Controller\\EventController->newAction(0, '')
#14 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(158): Mautic\\CoreBundle\\Controller\\CommonController->executeAction('new', 0, 0, '')
#15 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(80): Symfony\\Component\\HttpKernel\\HttpKernel->handleRaw(Object(Symfony\\Component\\HttpFoundation\\Request), 1)
#16 /var/www/html/vendor/symfony/http-kernel/Kernel.php(201): Symfony\\Component\\HttpKernel\\HttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#17 /var/www/html/app/AppKernel.php(115): Symfony\\Component\\HttpKernel\\Kernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#18 /var/www/html/app/middlewares/CORSMiddleware.php(82): AppKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#19 /var/www/html/app/middlewares/CatchExceptionMiddleware.php(34): Mautic\\Middleware\\CORSMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#20 /var/www/html/app/middlewares/Dev/IpRestrictMiddleware.php(52): Mautic\\Middleware\\CatchExceptionMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#21 /var/www/html/app/middlewares/VersionCheckMiddleware.php(58): Mautic\\Middleware\\Dev\\IpRestrictMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#22 /var/www/html/app/middlewares/TrustMiddleware.php(42): Mautic\\Middleware\\VersionCheckMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#23 /var/www/html/vendor/stack/builder/src/Stack/StackedHttpKernel.php(23): Mautic\\Middleware\\TrustMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#24 /var/www/html/vendor/stack/run/src/Stack/run.php(13): Stack\\StackedHttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request))
#25 /var/www/html/index_dev.php(25): Stack\
un(Object(Stack\\StackedHttpKernel))
#26 {main}
"} {"hostname":"mautic-web","pid":3260}
```

So to date I have not been able to get it to successfully sync all contacts or to push contacts to Hubspot with the integration.

Also, no companies at all have been created, but that might be because the ones which have been synchronised don't have a company associated? Not sure.

Side note: This PR might also fix https://github.com/mautic/mautic/issues/11064.

**Comment by npracht** [comment]
@RCheesley thanks for testing.
>Constantly getting the tooltip stuck on the browser:

This is an issue that appears everywhere in the app (in campaign especially). Not related to plugin.

>Only 8 contacts are synchronised with the cron job:

What is the parameters of your command ? It takes not all contacts, but base on a contact modified date timeframe.

@kuzmany can you check other feedbacks ?

**Comment by RCheesley** [comment]
> What is the parameters of your command ? It takes not all contacts, but base on a contact modified date timeframe.

Pretty sure I just ran the command from the docs with --all but it was a week ago now 😖

**Comment by npracht** [comment]
Doc added: https://github.com/mautic/mautic-documentation/pull/333

**Comment by volha-pivavarchyk** [comment]
Works well

1. HubSpot was configured well for me.
image

2. The cron job worked well too. All 5 contacts were added.

3. A campaign  with the action "Push contact to integration" works well too.

4. Result
image

image

**Review by volha-pivavarchyk — APPROVED** [review]

**Comment by escopecz** [comment]
@RCheesley I'm not clear whether you approve or block merging this PR.
@npracht can you please resolve the conflict and create a duplicate PR for the 5.x branch, please?

**Comment by adiux** [comment]
Thanks for this. It looks like there is an error for people who have an existing key saved. 

If there is an existing key I get a 500er error when saving the new key. I looks like it still uses the existing key in the url:

```
"HUBSPOT REQUEST URL: GET https://api.hubapi.com/contacts/v2/properties/?hapikey=pat-eu1-*******"
}
```

and the existing key can not be deleted, as it is a readonly field.
Screen Shot 2022-10-13 at 10 03 14

**Comment by RCheesley** [comment]
@kuzmany could you take a look at the feedback re. when there is already a key present, and resolve the conflict? Would be awesome to get this into the next release so that we have it out there before it reaches EOL.

**Comment by sensalot** [comment]
Mautic 4.4.4.
I can't make any changes to the config popup of hubspot. Even cancel gives a 500 error in return

**Comment by sensalot** [comment]
Mautic log:
[2022-10-26 13:02:59] mautic.CRITICAL: Uncaught PHP Exception Symfony\Component\Form\Exception\LogicException: "A form with an empty name cannot have a parent form." at /var/www/html/mautic/vendor/symfony/form/Form.php line 284 {"exception":"[object] (Symfony\\Component\\Form\\Exception\\LogicException(code: 0): A form with an empty name cannot have a parent form. at /var/www/html/mautic/vendor/symfony/form/Form.php:284)"}

**Comment by sensalot** [comment]
is there a workaround? Like adding the new API key manually to the database? If so which database field?

**Comment by stevedrobinson** [comment]
Tested on a clean install and I was able to authenticate, but advancing pages on a long list of contact fields doesn't work. 
image


The ajax call returns `{"success":0}` and the log contains the following error:
```
[2022-10-28 14:50:09] mautic.ERROR: INTEGRATION ERROR: Hubspot - Client error: `GET https://api.hubapi.com/contacts/v2/properties/` resulted in a `401 Unauthorized` response: {"status":"error","message":"Authentication credentials not found. This API supports both API Key and OAuth 2.0 authenti (truncated...)  [] {"hostname":"localhost","pid":109693}
```

**Comment by stevedrobinson** [comment]
I don't know if this has anything to do with the bug I mentioned above, but I also found [this bug and fix](https://forum.mautic.org/t/hubspot-mapping-error/25711/7) with the Hubspot plugin outside of this PR.

**Comment by sensalot** [comment]
I tested this suggestion of @stevedrobinson 
https://github.com/mautic/mautic/pull/11470#issuecomment-1295929636
But this did not fix this problem. (Chrome and Firefox)

**Comment by gonzalo** [comment]
Hi all, hubspot API keys deprecation will come effective on November 30. Will be this fix ready before that day? How it will be available? Through a standard Mautic update?
Thanks for the hard work!
~(I'd really would like to help with testing but as far as I read at the first post it's impossible to me as I only have access to a production environment)~

Well it was easier than expected to create a test environment. I can confirm that plugin works with my fresh test environment and it was able to export 3000 contacts.

Congratulations. I'm waiting for the merge!

**Comment by kuzmany** [comment]
@adiux your need write new access token  to `Hubspot Access token from private app` field. Second is for old api key, and cannot be modified (it's readonly) because BC

I was able to setup also on instance with old key.

**Comment by sensalot** [comment]
> @adiux your need write new access token to `Hubspot Access token from private app` field. Second is for old api key, and cannot be modified (it's readonly) because BC
> 
> I was able to setup also on instance with old key.

@kuzmany Opening and closing the popup for the hubspot plugin results in a 500 error at my side. Are you able to do this without an error?

**Comment by gonzalo** [comment]
Hi, only 7 days left to get old api keys for hubspot deprecated.

There's a plan to have this issue solved before the 30th of november?

**Comment by RCheesley** [comment]
@gonzalo we just need to have @npracht @kuzmany respond to the feedback and resolve the conflicts with the PR.

Are you also getting the issues that others have reported regarding the 500 error?

When you test, please report back your results as explained here: https://contribute.mautic.org/contributing-to-mautic/tester#leaving-your-review otherwise Github (and our Contribution tracking) doesn't know about your test :)

**Comment by RCheesley** [comment]
@all-contributors please add @gonzalo for userTesting

**Comment by gonzalo** [comment]
@RCheesley I was able to test it without any problem in dev environment but unfortunately I cannot (or don't know how to) test it in our production environment which has our current api-key set.

**Comment by sensalot** [comment]
I have tested version 4.4.45 in a gitpod using this link https://gitpod.io/#https://github.com/mautic/mautic and was able to edit and save the fields of the popup of the hubspot plugin.
But after reopening the message shows "The API key provided is invalid. View or manage your API key here: https://app.hubspot.com/l/api-key/". I have use the live key of a private app I created before (https://app-eu1.hubspot.com/private-apps/xxxxx)
As a result field mapping could not be tested.

**Comment by sensalot** [comment]
I have tested the gitpod of the pullrequest which shows version 4.4.41 in the simple browser in gitpod.
I was able to edit and save the fields. The API key was accepted and fieldmapping became available. I was able to map a contact field and a company field. I did not test if data passed through

**Comment by RCheesley** [comment]
> I cannot (or don't know how to) test it in our production environment which has our current api-key set.

Well, if you can make a development/testing area, the way to test it would be taking a backup of the files and then replacing them with the ones that are changed in the PR: https://github.com/mautic/mautic/pull/11470/files

With each file there you can click the three dots menu and go to View File to see the entire file as modified by this PR:

screenshot-github com-2022 11 23-15_47_53

If you then press 'raw' you will see the full file without any of the GitHub user interface which you could copy and paste.

This of course assumes that you don't have any modifications to those files.

Then you would clear the cache, regenerate assets with `bin/console m:a:g` and test - but emphasising that you should not use a live active environment for this kind of testing! :) 

Alternatively, you could follow these steps:

1. Spin up a test environment in Gitpod with the default 4.4 branch and set up Hubspot - here is a link to [Gitpod on 4.4](https://gitpod.io/#https://github.com/mautic/mautic/tree/4.4)
2. Test Hubspot works OK with that environment and your API key
3. In the Gitpod window, on the left you will see a Github icon. Click that and then browse the PRs until you find this one (#11470) - right click on it and go to Checkout. This will then apply the PR. As we are in dev mode by default we should not need to clear the cache or regenerate assets (but you can if you want), you might need to force-reload the browser.
4. The Hubspot API key should be there from before - now try adding the new one, and see if it all works OK after you do that.

**Comment by kuzmany** [comment]
I was able to authorize even If I had setup old key.  Don't see issue

image

**Comment by kuzmany** [comment]
@RCheesley I fixed 

 > PHP Notice - Trying to access array offset on value of type null

It was not related issue to this PR

**Comment by sensalot** [comment]
I have added the 5 files to my production environment. The hubspot plugin pop-up shows "loading". This makes no difference https://forum.mautic.org/t/hubspot-mapping-error/25711/7 
Log shows mautic.ERROR: INTEGRATION ERROR: Hubspot - {"status":"error","message":"This hapikey doesn't exist.","correlationId":"
But I can't change it when the popup sticks at loading

**Comment by onairmarc** [comment]
@RCheesley has there been any movement on this? EOL is this coming Wednesday.

**Comment by RCheesley** [comment]
@adiux can you please re-test entering the new API key and see if you are still able to replicate the issue you saw, as this is blocking the PR at this time and others

**Review by RCheesley — APPROVED** [review]
Tested this by adding the old Hubspot key on mautic/mautic then pulling this PR and applying the new Private App. 

Sync continues to work, pushing from a campaign also works (have to switch to prod mode due to a warning when you add push to integration on a campaign) and contacts are pushed to Hubspot as expected.

+1 from me, LGTM!

**Comment by RCheesley** [comment]
@kuzmany @escopecz this needs to go out in 4.4.5 but it is failing on test coverage. Anything that can be done to get this over the line?

File: app/bundles/PluginBundle/Form/Type/IntegrationsListType.php [hunk 2]
```diff
@@ -99,7 +99,7 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                     'attr'  => [
                         'class' => 'integration-config-container',
                     ],
-                    'integration' => (isset($integrationObjects[$data['integration']])) ? $integrationObjects[$data['integration']] : null,
+                    'integration' => (isset($integrationObjects[$data['integration'] ?? null])) ? $integrationObjects[$data['integration']] : null,
```
Comment (escopecz) [inline app/bundles/PluginBundle/Form/Type/IntegrationsListType.php:102]: This change doesn't make any sense. Either use `isset` or `??`. Why both?

Comment (kuzmany) [reply app/bundles/PluginBundle/Form/Type/IntegrationsListType.php:102]: ?? is on `$data['integration']` isset is on all `$integrationObjects[$data['integration'] ?? null]`

Two different variable. Just notice show up during the working on this PR

Comment (escopecz) [reply app/bundles/PluginBundle/Form/Type/IntegrationsListType.php:102]: Missed that it's 2 different arrays. Thanks for the explanation!

**Comment by RCheesley** [comment]
OK as we require this PR to get the release out and it is time sensitive I have created #11767 to track creating the 5.x PR and adding the test coverage after the release is made.

**Comment by gonzalo** [comment]
Update applied to my production environment. I was able to set the new private app token and launch the synchronization with all the fields.

Worked perfectly

Thanks!

**Comment by sensalot** [comment]
Update applied. Still get a 500 error an any button of the popup screen. Log shows "A form with an empty name cannot have a parent form."

[2022-11-30 22:05:24] mautic.CRITICAL: Uncaught PHP Exception Symfony\Component\Form\Exception\LogicException: "A form with an empty name cannot have a parent form." at /var/www/html/mautic/vendor/symfony/form/Form.php line 284 {"exception":"[object] (Symfony\\Component\\Form\\Exception\\LogicException(code: 0): A form with an empty name cannot have a parent form. at /var/www/html/mautic/vendor/symfony/form/Form.php:284)"} {"hostname":"37-97-186-212.colo.transip.net","pid":21447}

**Comment by sensalot** [comment]
System info showed some recommendations we followed by adding
date.timezone settint is not set in PHP config
imap extension
php_posix
intl extension

Now I don't get a 500 error. But it hangs on loading. Nothing in the log.

**Comment by RCheesley** [comment]
@sensalot please take this to the forums rather than on a closed PR. This is now in the 4.4.5 release so recommend you update to the new release if you haven't already done so. Thanks!

**Pull request merged**